### PR TITLE
[ADD] Improve Reconciliation Process - Part 1

### PR DESCRIPTION
**Pull request by smetl**
1) Payment communication
2) Replace old manual reconciliation wizard by the new one

task: https://www.odoo.com/web#id=1868491&action=333&active_id=967&model=project.task&view_type=form&menu_id=4720

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/l10n_be_invoice_bba/models/account_invoice.py [hunk 1]
```diff
@@ -20,14 +20,14 @@ class AccountInvoice(models.Model):
     _inherit = 'account.invoice'
 
     @api.model
-    def _get_reference_type(self):
+    def _get_reference_types(self):
         """Add BBA Structured Communication Type and change labels from 'reference' into 'communication' """
-        res = super(AccountInvoice, self)._get_reference_type()
+        res = super(AccountInvoice, self)._get_reference_types()
         res[[i for i, x in enumerate(res) if x[0] == 'none'][0]] = ('none', _('Free Communication'))
```
Comment (csnauwaert) [inline addons/l10n_be_invoice_bba/models/account_invoice.py]: It seems that this line is not necessary anymore as you've already rename the 'none' selection to 'Free Communication' in account_invoice.py

File: addons/account/views/res_config_settings_views.xml [hunk 2]
```diff
@@ -257,6 +257,20 @@
                                     </div>
                                 </div>
                             </div>
+                            <div class="col-xs-12 col-md-6 o_setting_box">
+                                <div class="o_setting_left_pane"/>
+                                <div class="o_setting_right_pane">
+                                    <label for="invoice_reference_type"/>
+                                    <div class="text-muted">
+                                        Ask communication on invoice reports.
```
Comment (csnauwaert) [inline addons/account/views/res_config_settings_views.xml]: Maybe 'Show communication on invoice reports' ?

File: addons/account/views/account_invoice_view.xml [hunk 3]
```diff
@@ -431,6 +431,9 @@
                             <field name="date_invoice"/>
                             <field name="date_due" attrs="{'readonly': ['|',('payment_term_id','!=',False), ('state', 'in', ['open', 'paid'])]}" force_save="1"/>
                             <field name="move_name" invisible="1"/>
+                            <field name="reference_type" nolabel="1"  attrs="{'readonly':[('state','!=','draft')]}"/>
+                            <field name="reference" nolabel="1"  attrs="{'readonly':[('state','!=','draft')]}"/>
+                            <field name="reference"/>
```
Comment (csnauwaert) [inline addons/account/views/account_invoice_view.xml]: Why displaying twice the reference field? Once in readonly if state is draft and the other always editable. Seems strange to me

File: addons/account/models/account_invoice.py [hunk 4]
```diff
@@ -388,6 +388,17 @@ def _get_vendor_display_info(self):
...
             invoice.invoice_icon = invoice.source_email and '@' or ''
 
+    @api.multi
+    def _get_computed_reference(self):
+        if self.reference_type == 'invoice_number':
+            trailing_digits_str = re.match('.*?([0-9]+)$', self.number).group(1)
+            digits = len(trailing_digits_str)
+            trailing_digits = int(trailing_digits_str)
+            magical_modulo = 97
+            return '%s-%s' % (self.number, str(trailing_digits % magical_modulo).rjust(digits, '0'))
```
Comment (csnauwaert) [inline addons/account/models/account_invoice.py]: I don't understand the use case: when would you want a communication that looks like: INV/2018/0001-0001? The last part will always be the same (we can't have INV/2018/0001-0002 for example), so why not use self.number instead? 
Edit: well actually we can have INV/2018/0098-0001 but I still fail to see how useful it is

File: addons/account/models/reconciliation_widget.py [hunk 5]
```diff
@@ -265,6 +265,7 @@ def get_data_for_manual_reconciliation(self, res_type, res_ids=None, account_typ
         assert account_type in ('payable', 'receivable', None)
         is_partner = res_type == 'partner'
         res_alias = is_partner and 'p' or 'a'
+        aml_ids = self._context.get('active_ids') and tuple(self._context.get('active_ids'))
```
Comment (csnauwaert) [inline addons/account/models/reconciliation_widget.py]: shouldn't we also check context.get('active_model') to be sure that the active_ids are coming from account_move_line and not from another source?

File: addons/account/wizard/account_reconcile_view.xml [hunk 6]
```diff
@@ -26,14 +26,11 @@
...
             <field name="name">Reconcile Entries</field>
-            <field name="res_model">account.move.line.reconcile</field>
-            <field name="view_type">form</field>
-            <field name="view_mode">form</field>
-            <field name="view_id" ref="view_account_move_line_reconcile_full"/>
-            <field name="target">new</field>
-            <field name="binding_model_id" ref="account.model_account_move_line" />
+            <field name="tag">manual_reconciliation_view</field>
+            <field name="binding_model_id" ref="account.model_account_move_line"/>
+            <field name="binding_type">action</field>
```
Comment (csnauwaert) [inline addons/account/wizard/account_reconcile_view.xml]: just by curiosity, what does it do? binding_*

**Comment by qdp-odoo** [comment]
merged in master, for saas-11.5, at revisions
https://github.com/odoo/odoo/commit/b5bb5bd421ad5c643f72602f6a0f97a4a0eb8711
https://github.com/odoo/odoo/commit/3cee86e1499e132ec6dbf1f14873a8f33114ecfa

File: addons/account/models/account_invoice.py (hunk 4 above)
Comment (nhomar) [reply addons/account/models/account_invoice.py]: @csnauwaert when you pay partially you will have several payments for the same invoice, that's what I get here, a perfect use case if I understand well enough.

File: addons/account/models/company.py [hunk 7]
```diff
@@ -66,6 +69,8 @@ class ResCompany(models.Model):
     account_setup_fy_data_done = fields.Boolean('Financial Year Setup Marked As Done', help="Technical field holding the status of the financial year setup step.")
     account_setup_coa_done = fields.Boolean(string='Chart of Account Checked', help="Technical field holding the status of the chart of account setup step.")
     account_setup_bar_closed = fields.Boolean(string='Setup Bar Closed', help="Technical field set to True when setup bar has been closed by the user.")
+    invoice_reference_type = fields.Selection(string='Default Invoice Reference Type', selection='_get_invoice_reference_types',
+                                              default='none', help='Default Reference Type on Invoices.')
```
Comment (nhomar) [inline addons/account/models/company.py]: this is an error use a method to fill the types, when we have the add option available in the new api don's you think?

**Pull request closed without merging**